contracts/governance/Governor.sol

```diff
@@ -274,17 +278,23 @@ abstract contract Governor is Context, ERC165, EIP712, IGovernor, IERC721Receive
         address proposer = _msgSender();
         uint256 currentTimepoint = clock();
 
-        require(
-            getVotes(proposer, currentTimepoint - 1) >= proposalThreshold(),
-            "Governor: proposer votes below proposal threshold"
-        );
+        // Avoid stack too deep
+        {
+            uint256 proposerVotes = getVotes(proposer, currentTimepoint - 1);
+            uint256 votesThreshold = proposalThreshold();
+            if (proposerVotes < votesThreshold) {
+                revert GovernorProposerInvalidTreshold(proposer, proposerVotes, votesThreshold);
+            }
+        }
 
         uint256 proposalId = hashProposal(targets, values, calldatas, keccak256(bytes(description)));
 
-        require(targets.length == values.length, "Governor: invalid proposal length");
-        require(targets.length == calldatas.length, "Governor: invalid proposal length");
-        require(targets.length > 0, "Governor: empty proposal");
-        require(_proposals[proposalId].voteStart == 0, "Governor: proposal already exists");
+        if (targets.length != values.length || targets.length != calldatas.length || targets.length == 0) {
+            revert GovernorInvalidProposalLength(targets.length, calldatas.length, values.length);
+        }
+        if (_proposals[proposalId].voteStart != 0) {
+            revert GovernorIncorrectState(proposalId, state(proposalId), bytes32(0));
```

Comment:
I'll go for `Unexpected`, I think it's a better description than `Incorrect` or `Unsupported` since the former indicates a problem with the state itself, and the later sounds like there's a state we don't support (which we do, the functions are just _expecting it_).

> Also how would you feel about using ProposalState instead of State? Perhaps it's redundant though.

It's more precise. Adding it.